File: test/csharp/SpecGate.CtscFixtures/Conformance/SumTypes/EnumEvent.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.SumTypes;

[SpecEvent("Shape")]
public abstract class Shape;

[SpecEvent("Circle")]
public sealed class Circle : Shape
{
    [SpecEvent("radius")]
    public double Radius { get; set; }
}

[SpecEvent("Rectangle")]
public sealed class Rectangle : Shape
{
    [SpecEvent("width")]
    public double Width { get; set; }

    [SpecEvent("height")]
    public double Height { get; set; }
}

[SpecEvent("Tag")]
public sealed class Tag : Shape
{
    [SpecEvent("value")]
    public string Value { get; set; } = string.Empty;
}

[SpecEvent("Point")]
public sealed class Point : Shape;

public static class EnumEvent
{
    [SpecOperation("classify", Spec = "fixture.enum_event")]
    public static Shape Classify([SpecInput("sides")] int sides) =>
        sides switch
        {
            0 => new Point(),
            1 => new Circle { Radius = 5.0 },
            3 => new Tag { Value = "triangle" },
            4 => new Rectangle { Width = 3.0, Height = 4.0 },
            _ => new Point(),
        };
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/sum_types/enum_event.rs
use specgate::{SpecEvent, spec_operation};

#[derive(Debug, Clone, PartialEq, SpecEvent)]
#[spec_component("fixture.enum_event")]
pub enum Shape {
    Circle { radius: f64 },
    Rectangle { width: f64, height: f64 },
    Tag(String),
    Point,
}

#[spec_operation("classify", spec = "fixture.enum_event")]
pub fn classify(sides: i32) -> Shape {
    match sides {
        0 => Shape::Point,
        1 => Shape::Circle { radius: 5.0 },
        3 => Shape::Tag("triangle".to_string()),
        4 => Shape::Rectangle {
            width: 3.0,
            height: 4.0,
        },
        _ => Shape::Point,
    }
}

#[test]
fn tagged_union_supports_unit_tuple_and_named_variants() {
    assert_eq!(classify(0), Shape::Point);
    assert_eq!(classify(3), Shape::Tag("triangle".to_string()));
    assert_eq!(
        classify(4),
        Shape::Rectangle {
            width: 3.0,
            height: 4.0
        }
    );
}
